File: api/download.js
// api/download.js
// Vercel Node.js serverless function
// POST { url, quality:"best|720|480|360", format:"video|mp3" }
// GET  ?url=...&quality=...&format=...

const { execFile } = require("child_process");
const path = require("path");

// yt-dlp-exec bundles yt-dlp binary — works on Vercel
let ytDlp;
try {
  ytDlp = require("yt-dlp-exec");
} catch (e) {
  ytDlp = null;
}

function cors(res) {
  res.setHeader("Access-Control-Allow-Origin",  "*");
  res.setHeader("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
  res.setHeader("Access-Control-Allow-Headers", "Content-Type, Accept");
  res.setHeader("Content-Type", "application/json");
}

function fmtSelector(quality, isAudio) {
  if (isAudio) return "bestaudio[ext=m4a]/bestaudio/best";
  const h = { best: null, "720": "720", "480": "480", "360": "360" }[quality] ?? "720";
  if (!h) return "bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best";
  return `bestvideo[height<=${h}][ext=mp4]+bestaudio[ext=m4a]/best[height<=${h}][ext=mp4]/best[height<=${h}]`;
}

async function getInfo(url, quality, isAudio) {
  if (!ytDlp) throw new Error("yt-dlp-exec not installed");

  const info = await ytDlp(url, {
    dumpSingleJson: true,
    noPlaylist: true,
    noWarnings: true,
    noCheckCertificates: true,
    noCacheDir: true,
    format: fmtSelector(quality, isAudio),
  });

  // grab direct stream url
  let dlUrl = info.url || "";
  if (!dlUrl && Array.isArray(info.requested_formats)) {
    for (const f of info.requested_formats) {
      if (f.url) { dlUrl = f.url; break; }
    }
  }
  if (!dlUrl) throw new Error("No playable URL returned");

  return {
    success:  true,
    url:      dlUrl,
    title:    info.title    || "Video",
    thumb:    info.thumbnail|| "",
    duration: info.duration || 0,
    ext:      isAudio ? "mp3" : (info.ext || "mp4"),
    filesize: info.filesize || info.filesize_approx || 0,
  };
}

module.exports = async function handler(req, res) {
  cors(res);

  // preflight
  if (req.method === "OPTIONS") {
    return res.status(204).end();
  }

  // parse params
  let url = "", quality = "best", format = "video";
  if (req.method === "POST") {
    const body = req.body || {};
    url     = (body.url     || "").trim();
    quality = (body.quality || "best").trim();
    format  = (body.format  || "video").trim();
  } else {
    const q = req.query || {};
    url     = (q.url     || "").trim();
    quality = (q.quality || "best").trim();
    format  = (q.format  || "video").trim();
  }

  if (!url || !url.startsWith("http")) {
    return res.status(400).json({ success: false, error: "url required" });
  }

  try {
    const result = await getInfo(url, quality, format === "mp3");
    return res.status(200).json(result);
  } catch (err) {
    return res.status(200).json({ success: false, error: String(err.message || err) });
  }
};

File: api/download.py
import json
import subprocess
import os
from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs


def _cors(h):
    h["Access-Control-Allow-Origin"]  = "*"
    h["Access-Control-Allow-Methods"] = "GET, POST, OPTIONS"
    h["Access-Control-Allow-Headers"] = "Content-Type"


def _ytdlp():
    for p in ["/var/task/bin/yt-dlp", "/tmp/yt-dlp", "/usr/local/bin/yt-dlp"]:
        if os.path.isfile(p):
            return p
    return "yt-dlp"


def _sel(quality, audio):
    if audio:
        return "bestaudio[ext=m4a]/bestaudio/best"
    h = {"best": None, "720": "720", "480": "480", "360": "360"}.get(quality, "720")
    if h is None:
        return "bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best"
    return (
        f"bestvideo[height<={h}][ext=mp4]+bestaudio[ext=m4a]"
        f"/best[height<={h}][ext=mp4]/best[height<={h}]"
    )


def _run(url, quality, audio):
    p = subprocess.run(
        [_ytdlp(), "--dump-json", "--no-playlist", "--no-warnings",
         "--no-check-certificates", "--no-cache-dir",
         "-f", _sel(quality, audio), url],
        capture_output=True, text=True, timeout=25
    )
    if p.returncode != 0:
        lines = (p.stderr or "").strip().splitlines()
        raise RuntimeError(lines[-1] if lines else "yt-dlp failed")

    d = json.loads(p.stdout)
    dl = d.get("url", "")
    if not dl:
        for f in d.get("requested_formats", []):
            if f.get("url"):
                dl = f["url"]; break
    if not dl:
        raise RuntimeError("No playable URL")

    return {
        "success":  True,
        "url":      dl,
        "title":    d.get("title", "Video"),
        "thumb":    d.get("thumbnail", ""),
        "duration": d.get("duration", 0),
        "ext":      "mp3" if audio else d.get("ext", "mp4"),
        "filesize": d.get("filesize") or d.get("filesize_approx") or 0,
    }


class handler(BaseHTTPRequestHandler):

    def log_message(self, *a): pass

    def _send(self, code, data):
        body = json.dumps(data).encode()
        self.send_response(code)
        _cors(self.headers)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def do_OPTIONS(self):
        self.send_response(204)
        _cors(self.headers)
        self.end_headers()

    def do_GET(self):
        qs  = parse_qs(urlparse(self.path).query)
        url = qs.get("url",     [""])[0].strip()
        q   = qs.get("quality", ["best"])[0].strip()
        fmt = qs.get("format",  ["video"])[0].strip()
        self._go(url, q, fmt)

    def do_POST(self):
        n = int(self.headers.get("Content-Length", 0))
        try:    body = json.loads(self.rfile.read(n))
        except: body = {}
        url = body.get("url",     "").strip()
        q   = body.get("quality", "best").strip()
        fmt = body.get("format",  "video").strip()
        self._go(url, q, fmt)

    def _go(self, url, q, fmt):
        if not url or not url.startswith("http"):
            self._send(400, {"success": False, "error": "url required"}); return
        try:
            self._send(200, _run(url, q, fmt == "mp3"))
        except Exception as e:
            self._send(200, {"success": False, "error": str(e)})
